fastjsonschema/generator.py

```diff
@@ -271,29 +274,43 @@ def generate_not(self):
 
             {'not': {'type': 'null'}}
 
-        Valid values for this definitions are 'hello', 42, ... but not None.
+        Valid values for this definitions are 'hello', 42, {} ... but not None.
         """
-        with self.l('try:'):
-            self.generate_func_code_block(self._definition['not'], self._variable, self._variable_name)
-        self.l('except JsonSchemaException: pass')
-        self.l('else: raise JsonSchemaException("{name} must not be valid by not definition")')
+        if self._definition['not'] is not None and not self._definition['not']:  # {}
+            with self.l('if "{}" in {}.keys():', self._context['key'], self._context['definition'].get('properties', {})):
+                self.l('raise JsonSchemaException("{name} must not be valid by not definition")')
+        else:
+            with self.l('try:'):
+                self.generate_func_code_block(self._definition['not'], self._variable, self._variable_name)
+            self.l('except JsonSchemaException: pass')
+            self.l('else: raise JsonSchemaException("{name} must not be valid by not definition")')
 
     def generate_min_length(self):
+        with self.l('if not isinstance({variable}, str):'):
+            self.l('return {variable}')
         self.create_variable_with_length()
         with self.l('if {variable}_len < {minLength}:'):
             self.l('raise JsonSchemaException("{name} must be longer than or equal to {minLength} characters")')
 
     def generate_max_length(self):
+        with self.l('if not isinstance({variable}, str):'):
+            self.l('return {variable}')
         self.create_variable_with_length()
         with self.l('if {variable}_len > {maxLength}:'):
             self.l('raise JsonSchemaException("{name} must be shorter than or equal to {maxLength} characters")')
 
     def generate_pattern(self):
+        with self.l('if not isinstance({variable}, str):'):
+            self.l('return {variable}')
         self._compile_regexps['{}_re'.format(self._variable)] = re.compile(self._definition['pattern'])
-        with self.l('if not {variable}_re.match({variable}):'):
+        with self.l('if not {variable}_re.search({variable}):'):
```

Comment:
Why this change? :-)

Reply:
> When defining the regular expressions, it’s important to note that the string is considered valid if the expression matches anywhere within the string.

– https://spacetelescope.github.io/understanding-json-schema/reference/string.html#index-2

> Python offers two different primitive operations based on regular expressions: re.match() checks for a match only at the beginning of the string, while re.search() checks for a match anywhere in the string

– http://devdocs.io/python~3.6/library/re#search-vs-match

Reply:
I see! I misinterpreted the validity of regexp in JSON schema. Thanks for the fix!